sdk/middleware/sqhttp/http.go

```diff
@@ -0,0 +1,34 @@
+package sqhttp
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/sqreen/go-agent/sdk"
+)
+
+// Middleware is Sqreen's middleware function for `net/http` to monitor and
+// protect received requests. It creates and stores the HTTP request record into
+// the request context so that it can be later accessed to perform SDK calls in
+// the decorated handler using `sdk.FromContext()`.
+//
+//	fn := func(w http.ResponseWriter, r *http.Request) {
+//		sdk.FromContext(r.Context()).TrackEvent("my.event")
+//		fmt.Fprintf(w, "OK")
+//	}
+//	http.HandleFunc("/foo", sqhttp.Middleware(http.HandlerFunc(fn)))
+//
+func Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Create a new request record for this request.
+		sqreen := sdk.NewHTTPRequestRecord(r)
+		defer sqreen.Close()
+
+		// Store it into the request's context.
+		ctx := r.Context()
+		contextKey := sdk.HTTPRequestRecordContextKey.String
+		ctx = context.WithValue(ctx, contextKey, sqreen)
+		r = r.WithContext(ctx)
+		next.ServeHTTP(w, r)
```

Comment:
Hah, I was used to just call `next(f, w)` when writing `HandlerFunc`s so was surprised by that one. Turns out `ServeHTTP` just does that:

```
// ServeHTTP calls f(w, r).
  1963  func (f HandlerFunc) ServeHTTP(w ResponseWriter, r *Request) {
  1964  	f(w, r)
  1965  }
```

Weird.

Reply:
They are equivalent, depending on the type:
var f http.HandlerFunc --> f(w, r)
var h http.Handler --> h.ServeHTTP(w, r)